### PR TITLE
V15 Standalone - Support taking snapshots

### PR DESCRIPTION
**Pull request by joshuahuburn**
#### Description
Added support for taking snapshots with the standalone runtime. They now get dumped into the `spatial/snapshots` folder. Might add a configure for this later or stick it in a deployment unique folder... Considering.

#### Tests
Tests pass 
image

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 1]
```diff
@@ -199,7 +199,10 @@ void FSpatialGDKEditorToolbarModule::PreUnloadCallback()
 
 	if (AutoStopLocalDeployment != EAutoStopLocalDeploymentMode::Never)
 	{
-		InspectorProcess->Cancel();
+		if (InspectorProcess && InspectorProcess->Update())
+		{
+			InspectorProcess->Cancel();
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp:204]: Driveby fix. This was my b, we never checked if the process existed.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 2]
```diff
@@ -240,9 +240,14 @@ void FLocalDeploymentManager::TryStartLocalDeployment(FString LaunchConfig, FStr
...
 	FString RuntimeArgs = FString::Printf(TEXT("--config=\"%s\" --snapshot=\"%s\" --worker-port %s --http-port %s --grpc-port %s "
-											   "--worker-external-host %s --snapshots-directory=\"%s\" %s"),
+											   "--snapshots-directory=\"%s\" %s"),
 										  *LaunchConfig, *SnapshotName, *FString::FromInt(WorkerPort), *FString::FromInt(HTTPPort),
-										  *FString::FromInt(GRPCPort), *RuntimeIPToExpose, *SnapshotPath, *LaunchArgs);
+										  *FString::FromInt(GRPCPort), *SnapshotPath, *LaunchArgs);
+
+	if (!RuntimeIPToExpose.IsEmpty())
+	{
+		RuntimeArgs.Append(FString::Printf(TEXT(" --worker-external-host %s"), *RuntimeIPToExpose));
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:249]: Another fix for earlier mistakes. The snapshot-directory was not getting parsed correctly if we didn't have a RuntimeIPToExpose set.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 3]
```diff
@@ -361,113 +366,61 @@ void FLocalDeploymentManager::SetAutoDeploy(bool bInAutoDeploy)
 }
 
 // TODO: UNR-4423 Change this to taking a snapshot without the services
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:368]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 4]
```diff
@@ -361,113 +366,61 @@ void FLocalDeploymentManager::SetAutoDeploy(bool bInAutoDeploy)
...
+			if (OnSnapshotTaken != nullptr)
 			{
-				UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Failed to trigger snapshot at '%s'; received '%s'"),
-					   *HttpRequest->GetURL(), *HttpResponse->GetContentAsString());
+				OnSnapshotTaken(false /* bSuccess */, FString() /* PathToSnapshot */);
+			}
+			return;
+		}
+
+		// Unfortunately by the time this callback happens, the files haven't been flushed, so if you copy you may get
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:387]: This should no longer be needed, right?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 5]
```diff
@@ -361,113 +366,61 @@ void FLocalDeploymentManager::SetAutoDeploy(bool bInAutoDeploy)
...
+					OnSnapshotTaken(bSuccess, NewestSnapshotFilePath);
 				}
-				return;
-			}
+			},
+			0.5f /* InRate */, false /* InbLoop */);
+	});
+
+	HttpRequest->SetURL(TEXT("http://localhost:5006/snapshot"));
+	HttpRequest->SetHeader("Content-Type", TEXT("application/json"));
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:421]: Probably not required, no content section in the request?

**Review by improbable-danny — APPROVED** [review]
👍

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 4 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:386]: I'll try without, fair point. Not sure on the semantics of the new snapshotting

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:386]: Confirmed with runtime, you're correct. Removing the timer.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 5 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:420]: Correct!

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 6]
```diff
@@ -233,16 +233,30 @@ void FLocalDeploymentManager::TryStartLocalDeployment(FString LaunchConfig, FStr
...
+
+	// Create the folder for storing the snapshots.
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	PlatformFile.CreateDirectoryTree(*SnapshotPath);
+
+	// runtime.exe --config=squid_config.json --snapshot=snapshots/default.snapshot --worker-port 8018 --http-port 5006 --grpc-port 7777
+	// --worker-external-host 127.0.0.1 --snapshots-directory=spatial/snapshots --schema-bundle=spatial/build/assembly/schema/schema.sb
 	FString RuntimeArgs = FString::Printf(TEXT("--config=\"%s\" --snapshot=\"%s\" --worker-port %s --http-port %s --grpc-port %s "
-											   "--worker-external-host %s --snapshots-directory=\"%s\" %s"),
+											   "--snapshots-directory=\"%s\" --schema-bundle=\"%s\" %s"),
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:252]: We now work with 15.0.0-preview-3 so we need this.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 7]
```diff
@@ -233,16 +233,30 @@ void FLocalDeploymentManager::TryStartLocalDeployment(FString LaunchConfig, FStr
...
+	// Give the snapshot path a timestamp to ensure we don't overwrite snapshots from older deployments.
+	// The snapshot service saves snapshots with the name `snapshot-n.snapshot` for a given deployment,
+	// where 'n' is the number of snapshots taken since starting the deployment.
+	// TODO: Change the TimeNow to be some arg included with the TryStartLocalDeployment which is also used for the logging folder.
+	FDateTime TimeNow = FDateTime::Now();
+	FString SnapshotPath = FPaths::Combine(SpatialGDKServicesConstants::SpatialOSSnapshotFolderPath, *TimeNow.ToString());
+
+	// Create the folder for storing the snapshots.
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	PlatformFile.CreateDirectoryTree(*SnapshotPath);
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:247]: do we need to handle file io errors or does CreateDirectoryTree do that for us?
